Add config field to specify chroot mapping for exec driver

parasyte opened the pull request:
- Same format as used by the internal chroot mapping
- Map: source_path -> dest_path
- Example HCL:

```
client {
  chroot_env {
    "/etc" = "/etc"
    "/lib" = "/lib"
    "/opt/projects/foo/bin" = "/usr/bin"
  }
}
```

parasyte commented:
This can address #1507 by specifying a smaller subset of files and directories for your chroot.

diptanu commented on client/driver/executor/executor_linux.go:
```diff
@@ -227,7 +227,12 @@ func (e *UniversalExecutor) configureChroot() error {
 
-	if err := allocDir.Embed(e.ctx.Task.Name, chrootEnv); err != nil {
+	chroot := chrootEnv
+	if e.command.ChrootEnv != nil && len(e.command.ChrootEnv) > 0 {
```
`if len(e.command.ChrootEnv) > 0 ` since `len` of `nil` maps are `0`

diptanu commented on command/agent/config_parse.go:
```diff
@@ -370,6 +372,20 @@ func parseClient(result **ClientConfig, list *ast.ObjectList) error {
 
+	// Parse out chroot_env fields. These are in HCL as a list so we need to
+	// iterate over them and merge them.
+	if chrootEnvO := listVal.Filter("chroot_env"); len(chrootEnvO.Items) > 0 {
```
Please add a test for this

diptanu commented on client/driver/executor/executor.go:
```diff
@@ -126,6 +126,10 @@ type ExecCommand struct {
 
+	// A mapping of directories on the host OS to attempt to embed inside each
+	// task's chroot.
+	ChrootEnv map[string]string
```
Please move this field to `ExecutorContext`

parasyte commented:
@diptanu Addressed all review comments. I haven't been able to validate the new tests, because the test suite is failing on the point in master where I forked.

dadgar commented:
Can you also add documentation to the website /docs/agent/config.html

parasyte commented:
@dadgar :+1: Incoming.

dadgar commented:
Thanks for the docs they look great. Though the test is failing: https://travis-ci.org/hashicorp/nomad/builds/150180275#L3656

parasyte commented:
Yep, it's a bad test. I know `alloc`, `dev`, and `proc` are mounted into the chroot (forgot about those), but where do `tmp` and `local` come from? Also, I should really look into running individual tests to speed up the dev cycles.

jshaw86 commented:
@dadgar @diptanu what milestone will you all be bringing this PR into? Trying to decide if we need a build pipeline internally for a while till it lands into a milestone

diptanu commented:
@parasyte `tmp` and `local` are created by the nomad client in every allocation directory.

diptanu commented:
@jshaw86 I am hoping this would get into the 0.4.1 release if everything goes well.

parasyte commented:
@dadgar The test should be fixed now!

@diptanu Thanks, I was able to find where the local directories were created.

dadgar commented:
This test is also failing: https://travis-ci.org/hashicorp/nomad/builds/150759498#L5510

parasyte commented:
Ok, got that one too!

diptanu commented:
@parasyte Thanks for the contribution! This has landed on master now.

cc/ @jshaw86
